### PR TITLE
Feature/torus

### PR DESCRIPTION
**Pull request by jackraymond**
The primary purpose of this pull request is to provide generators for the instances evaluated in https://arxiv.org/abs/2202.03044 [BBRR]. 

The pull request:
- Adds support for torus graph variations upon Chimera, Pegasus and Zephyr as used in _[BBRR], _[RH].
- Corrects errors associated to special edge cases in use of check_edge_list and check_node_list in zephyr_graph, pegasus_graph and chimera_graph
- Refactors some bibliography entries, and corrects for some associated errors in elimination_order.py
- Simplifies one line in coloring.py to suppress a type error.

### CODE REVIEW

File: dwave_networkx/generators/chimera.py [hunk 1]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+        a warning is issued and the values are ignored.
+
+    Returns
+    -------
+    G : NetworkX Graph
+        A Chimera torus with shape (m, n, t), with Chimera coordinate node labels.
+
+
+    A Chimera torus is a generalization of the standard chimera graph
+    whereby bulk connectivity properties are maintained, but the boundary
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:777]:
```suggestion
    whereby degree-six connectivity is maintained, but the boundary
```

File: dwave_networkx/generators/chimera.py [hunk 2]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+    condition is modified to enforce an additional translational 
+    invariance symmetry [RH]_. Local connectivity in the chimera_torus
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:779]:
```suggestion
    condition is modified to enforce an additional translational-invariance 
    symmetry [RH]_. Local connectivity in the Chimera torus
```

File: docs/bibliography.rst [hunk 3]
```diff
@@ -10,3 +10,10 @@ Bibliography
 .. [FIA] Facchetti, G., Iacono G., and Altafini C. (2011). Computing global structural balance in large-scale signed social networks. PNAS, 108, no. 52, 20953-20958
 
 .. [DWMP] Dahl, E., "Programming the D-Wave: Map Coloring Problem", https://www.dwavesys.com/media/htfgw5bk/map-coloring-wp2.pdf
+
+.. [BBRR] Boothby, K., P. Bunky, J. Raymond, A. Roy, (2019). Next-Generation Topology of D-Wave Quantum Processors. https://arxiv.org/abs/2003.00133
```
Comment (randomir) [inline docs/bibliography.rst:14]: Bunyk

File: dwave_networkx/generators/chimera.py [hunk 4]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+    is identical to connectivity for chimera graph nodes away from the boundary.
+    The graph has :code:`V=8*m*n` nodes, and :code:`min(6,4+m)V//2 + 
+    min(6,4+n)V/2` edges. With the standard :math:`K_{t,t}` Chimera tile definition, 
+    any (x rows,y columns) tile displacement modulo (m,n)
+    :code:`(i,j,u,k)` -> :code:`((i+x)%m,(i+y)%n,u,k)`
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:784]:
```suggestion
    any tile displacement :math:`(x, y)`  modulo :math:`(m,n)`, rows and columns respectively,
    that is, :code:`(i,j,u,k)` -> :code:`((i+x)%m,(i+y)%n,u,k)`,
```

File: dwave_networkx/generators/chimera.py [hunk 5]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+    defines an automorphism.
+
+    See ``chimera_graph()`` for additional information.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:787]:
```suggestion
    See :func:`.chimera_graph` for additional information.
```

File: dwave_networkx/generators/chimera.py [hunk 6]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
 
+def chimera_torus(m, n=None, t=None, node_list=None, edge_list=None):
+    """Creates a defect-free Chimera lattice of size (m, n, t) subject to periodic boundary conditions
+
+
+    Parameters
+    ----------
+    m : int
+        Number of rows in the Chimera torus lattice.
+        If m<3 translational invariance already applies in the rows. If
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:749]:
```suggestion
        If :math:`m<3` translational invariance already applies in the rows. If 
```

File: dwave_networkx/generators/chimera.py [hunk 7]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+def chimera_torus(m, n=None, t=None, node_list=None, edge_list=None):
+    """Creates a defect-free Chimera lattice of size (m, n, t) subject to periodic boundary conditions
+
+
+    Parameters
+    ----------
+    m : int
+        Number of rows in the Chimera torus lattice.
+        If m<3 translational invariance already applies in the rows. If 
+        m>=3 additional external couplers, establishing translational invariance.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:750]: "If m>=3 additional external couplers, establishing translational invariance." --> missing some verb

File: dwave_networkx/generators/chimera.py [hunk 8]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+        Connectivity of all horizontal qubits is :math:`min(m-1,2)+2t`.
+    n : int (optional, default m)
+        Number of columns in the Chimera torus lattice.
+        If n<3 translational invariance already applies in the columns. If
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:754]:
```suggestion
        If :math:`n<3` translational invariance already applies in the columns. If 
```

File: dwave_networkx/generators/chimera.py [hunk 9]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+    ----------
+    m : int
+        Number of rows in the Chimera torus lattice.
+        If m<3 translational invariance already applies in the rows. If 
+        m>=3 additional external couplers, establishing translational invariance.
+        Connectivity of all horizontal qubits is :math:`min(m-1,2)+2t`.
+    n : int (optional, default m)
+        Number of columns in the Chimera torus lattice.
+        If n<3 translational invariance already applies in the columns. If 
+        n>=3 additional external couplers, establishing translational invaraince.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:755]: Same as the previous comment, missing some verb

File: dwave_networkx/generators/chimera.py [hunk 10]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+        Connectivity of all vertical qubits is :math:`min(n-1,2)+2t`.
+    t : int (optional, default 4)
+        Size of the shore within each Chimera tile.
+    node_list : iterable (optional, default None)
+        Iterable of nodes in the graph. If None, nodes are
+        generated as described below. If values incompatible with
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:761]:
```suggestion
        generated as described for :func:`.chimera_graph`. If values are incompatible with
```

File: dwave_networkx/generators/chimera.py [hunk 11]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+        the integer (or coordinates, as appropriate) labeling are proposed 
+        a warning is issued and the values are ignored.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:763]:
```suggestion
        the labeling  (integer or coordinates, as appropriate) are proposed, 
        a warning is issued and the values are ignored.
```

File: dwave_networkx/generators/chimera.py [hunk 12]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+    edge_list : iterable (optional, default None)
+        Iterable of edges in the graph. If None, edges are
+        generated as described below. If values incompatible with
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:766]:
```suggestion
        generated as described for :func:`.chimera_graph`. If values are incompatible with 
```

File: dwave_networkx/generators/chimera.py [hunk 13]
```diff
@@ -726,3 +738,91 @@ def chimera_to_target(q):
...
+    node_list : iterable (optional, default None)
+        Iterable of nodes in the graph. If None, nodes are
+        generated as described below. If values incompatible with
+        the integer (or coordinates, as appropriate) labeling are proposed 
+        a warning is issued and the values are ignored.
+    edge_list : iterable (optional, default None)
+        Iterable of edges in the graph. If None, edges are
+        generated as described below. If values incompatible with 
+        the integer (or coordinates, as appropriate) labeling are proposed 
+        a warning is issued and the values are ignored.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/chimera.py:768]:
```suggestion
        the labeling  (integer or coordinates, as appropriate) are proposed, 
        a warning is issued and the values are ignored.
```

**Review by JoelPasvolsky — COMMENTED** [review]
I'm making this PR in slices to help you get it in quickly. This first one is mostly on the `chimera.py` updates

File: dwave_networkx/generators/pegasus.py [hunk 14]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+    -------
+    G : NetworkX Graph
+        A Pegasus torus for size parameter `m` using the coordinate labeling system.
+
+
+    A pegasus torus is a generalization of the standard pegaus graph
+    whereby bulk connectivity properties are maintained, but the boundary
+    condition is modified to enforce an additional translational 
+    invariance symmetry [RH]_. Local connectivity in the pegasus torus
+    is identical to connectivity for pegasus graph nodes away from the boundary.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1228]:
```suggestion
    is identical to connectivity for Pegasus graph nodes away from the boundary.
```

File: dwave_networkx/generators/pegasus.py [hunk 15]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+    A tile consists of 24 nodes, and the torus has m-1 by m-1 tiles.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1229]:
```suggestion
    A tile consists of 24 nodes, and the torus has :math:`m-1` by :math:`m-1` tiles. 
```

Note that we don't really define "tile", the [Pegasus description](https://docs.dwavesys.com/docs/latest/c_gs_4.html#pegasus-couplers) talks about "A Pegasus unit **cell** contains twenty-four qubits"

File: dwave_networkx/generators/pegasus.py [hunk 16]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+        A Pegasus torus for size parameter `m` using the coordinate labeling system.
+
+
+    A pegasus torus is a generalization of the standard pegaus graph
+    whereby bulk connectivity properties are maintained, but the boundary
+    condition is modified to enforce an additional translational 
+    invariance symmetry [RH]_. Local connectivity in the pegasus torus
+    is identical to connectivity for pegasus graph nodes away from the boundary.
+    A tile consists of 24 nodes, and the torus has m-1 by m-1 tiles. 
+    Tile displacement modulo m-1 defines an automorphism.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1230]:
```suggestion
    Tile displacement modulo :math:`m-1` defines an automorphism.
```

File: dwave_networkx/generators/pegasus.py [hunk 17]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+    
+    See ``pegasus_graph()`` for additional information.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1232]:
```suggestion
    See :func:`.pegasus_graph` for additional information.
```

File: dwave_networkx/generators/pegasus.py [hunk 18]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+    """
+    Creates a Pegasus graph modified to allow for periodic boundary conditions and translational invariance.
+
+    Parameters
+    ----------
+    m : int
+        Size parameter for the Pegasus lattice.
+        Connectivity of all nodes is :math:`13+min(m-1,2)`
+    node_list : iterable, optional (default None)
+        Iterable of nodes in the graph. If None, calculated from `m`.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1201]:
```suggestion
        Iterable of nodes in the graph. If None, calculated from :math:``m`.
```

File: dwave_networkx/generators/pegasus.py [hunk 19]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+        Note that this list is used to remove nodes, so any nodes specified
+        not in ``range(24 * m * (m-1))`` are not added.
+    edge_list : iterable, optional (default None)
+        Iterable of edges in the graph. If None, edges are generated as
+        described below. The nodes in each edge must be integer-labeled in
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1206]:
```suggestion
        described in :func:`.pegasus_graph`. The nodes in each edge must be integer-labeled in
```

File: dwave_networkx/generators/pegasus.py [hunk 20]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+        ``range(24 * m * (m-1))``.
+    offset_lists : pair of lists, optional (default None)
+        Directly controls the offsets. Each list in the pair must have length 12
+        and contain even ints.  If `offset_lists` is not None, the `offsets_index`
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1210]:
```suggestion
        and contain even integers.  If ``offset_lists`` is not None, the ``offsets_index``
```

File: dwave_networkx/generators/pegasus.py [hunk 21]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+        parameter must be None.
+    offsets_index : int, optional (default None)
+        A number between 0 and 7, inclusive, that selects a preconfigured
+        set of topological parameters. If both the `offsets_index` and
+        `offset_lists` parameters are None, the `offsets_index` parameters is set
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1215]:
```suggestion
        set of topological parameters. If both the ``offsets_index`` and
        ``offset_lists`` parameters are None, the ``offsets_index`` parameters is set
```

File: dwave_networkx/generators/pegasus.py [hunk 22]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+        to zero. At least one of these two parameters must be None.
+
+    Returns
+    -------
+    G : NetworkX Graph
+        A Pegasus torus for size parameter `m` using the coordinate labeling system.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/pegasus.py:1221]:
```suggestion
        A Pegasus torus for size parameter :math:`m` using the coordinate labeling system.
```

**Review by JoelPasvolsky — COMMENTED** [review]
Comment for another slice: `pegasus.py`

File: dwave_networkx/generators/pegasus.py [hunk 23]
```diff
@@ -1178,3 +1186,89 @@ def source_to_inner(q):
...
+
+
+    A pegasus torus is a generalization of the standard pegaus graph
```
[the remark this answers is not in the document]
Comment (randomir) [reply dwave_networkx/generators/pegasus.py:1224]: Pegasus :stuck_out_tongue:

[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dwave_networkx/generators/pegasus.py:1224]: Thank you sir!

File: dwave_networkx/generators/zephyr.py [hunk 24]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
+
+    A Zephyr torus is a generalization of the standard zephyr graph
+    whereby bulk connectivity properties are maintained, but the boundary
+    condition is modified to enforce an additional translational 
+    invariance symmetry [RH]_. Local connectivity in the zephyr torus
+    is identical to connectivity for zephyr graph nodes away from the boundary.
+    A tile consists of 8t nodes, and the torus has m by m tiles. 
+    Tile displacement modulo m defines an automorphism.
+    
+    See ``zephyr_graph()`` for additional information.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:732]:
```suggestion
    See :func:`.zephyr_graph` for additional information.
```

File: dwave_networkx/generators/zephyr.py [hunk 25]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
+    -------
+    G : NetworkX Graph
+        A Zephyr torus with grid parameter ``m`` and tile parameter ``t``,
+        with Zephyr coordinate node labels.
+
+
+    A Zephyr torus is a generalization of the standard zephyr graph
+    whereby bulk connectivity properties are maintained, but the boundary
+    condition is modified to enforce an additional translational 
+    invariance symmetry [RH]_. Local connectivity in the zephyr torus
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:727]:
```suggestion
    condition is modified to enforce an additional translational-invariance 
    symmetry [RH]_. Local connectivity in the Zephyr torus
```

File: dwave_networkx/generators/zephyr.py [hunk 26]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
+    is identical to connectivity for zephyr graph nodes away from the boundary.
+    A tile consists of 8t nodes, and the torus has m by m tiles. 
+    Tile displacement modulo m defines an automorphism.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:730]:
```suggestion
    Tile displacement modulo :math:`m` defines an automorphism.
```

File: dwave_networkx/generators/zephyr.py [hunk 27]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
     for offset in offset_list:
         yield make_mapping(source_to_inner, zephyr_to_target, offset)
+
+def zephyr_torus(m, t=4, node_list=None, edge_list=None):
+    """
+    Creates a Zephyr graph modified to allow for periodic boundary conditions and translational invariance.
+    
+    The graph matches the local connectivity properties of a standard zephyr graph,
+    but with modified periodic boundary condition. Tiles of 8t qubits are arranged
+    on an m by m torus.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:698]:
```suggestion
    The graph matches the local connectivity properties of a standard Zephyr graph,
    but with modified periodic boundary condition. Tiles of :math:`8t` nodes are arranged
    on an :math:`m` by :math:`m` torus. 
```

File: dwave_networkx/generators/zephyr.py [hunk 28]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
+    
+    The graph matches the local connectivity properties of a standard zephyr graph,
+    but with modified periodic boundary condition. Tiles of 8t qubits are arranged
+    on an m by m torus. 
+
+    Parameters
+    ----------
+    m : int
+        Grid parameter for the Zephyr lattice.
+        Connectivity of all nodes is :math:`4t+min(2m-1,4)`
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:704]:
```suggestion
        Connectivity of all nodes is :math:`4t+min(2m-1,4)`.
```

File: dwave_networkx/generators/zephyr.py [hunk 29]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
+    t : int
+        Tile parameter for the Zephyr lattice.
+    node_list : iterable, optional (default None)
+        Iterable of nodes in the graph. If None, calculated from ``m``.
+        Note that this list is used to remove nodes, so only specified nodes
+        that belong to the base node set (described in the ``coordinates``
+        parameter) are added.
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:711]:
```suggestion
        parameter of :func:`.zephyr_graph`) are added.
```

I added this before I saw line 715 (````coordinates`` parameter.    coordinates : bool, optional (default False)``) so ignore if not intended

File: dwave_networkx/generators/zephyr.py [hunk 30]
```diff
@@ -680,3 +688,85 @@ def source_to_inner(q):
...
+    edge_list : iterable, optional (default None)
+        Iterable of edges in the graph. If None, edges are generated as
+        described below. The nodes in each edge must be labeled according to the
+        ``coordinates`` parameter.    coordinates : bool, optional (default False)
```
Comment (JoelPasvolsky) [inline dwave_networkx/generators/zephyr.py:715]:
```suggestion
        ``coordinates`` parameter.    
coordinates : bool, optional (default False)
```

Not sure the formatting will work through a GitHub proposed change but the ``coordinates `` is currently appended to ``edge_list `` and missing a description

**Review by JoelPasvolsky — COMMENTED** [review]
Third slice (for Zephyr). I think that should cover most of what I'll spend time looking at

Very nice addition, @jackraymond

File: dwave_networkx/generators/zephyr.py (hunk 30 above)
Comment (jackraymond) [reply dwave_networkx/generators/zephyr.py:715]: I decided to remove possibility of coordinates option and just return coordinate vectors. I should amend the docstrings accordingly

**Comment by jackraymond** [comment]
Implemented review comments in full. Some docstring issues remained in edge_list and node_list, these have been subsequently simplified. Some other minor formatting and compliance issues were addressed.

**Review by randomir — APPROVED** [review]
Tests could use some PEP8 love, but otherwise LGTM!

File: docs/bibliography.rst [hunk 31]
```diff
@@ -10,3 +10,10 @@ Bibliography
 .. [FIA] Facchetti, G., Iacono G., and Altafini C. (2011). Computing global structural balance in large-scale signed social networks. PNAS, 108, no. 52, 20953-20958
 
 .. [DWMP] Dahl, E., "Programming the D-Wave: Map Coloring Problem", https://www.dwavesys.com/media/htfgw5bk/map-coloring-wp2.pdf
+
+.. [BBRR] Boothby, K., P. Bunyk, J. Raymond, A. Roy, (2019). Next-Generation Topology of D-Wave Quantum Processors. https://arxiv.org/abs/2003.00133
```
[the remark this answers is not in the document]
Comment (jackraymond) [reply docs/bibliography.rst]: I corrected several formatting inconsistencies in that file including that one.

**Pull request merged**